### PR TITLE
Add GSoC student report

### PR DESCRIPTION
**Pull request by harukamm**
about lambda support in debugger expression evaluator

### CODE REVIEW

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+My project is lambda support in the MonoDevelop debugger's expression evaluator, used in the Immediate pad.
+
+# Code
+
+All of my work is covered by a single pull request: [here](https://github.com/mono/debugger-libs/pull/142#pull)
+
+# Summary of the implementation
+
+![Processing flow chart](/images/gsoc-2017-lambda-support-in-debugger expression-evaluator.png "Processing flow chart")
```
Comment (marek-safar) [inline community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: You are missing dash in the file name

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+*Fig 1. Processing flow chart*
+
+The debugger works using an interface [SDWP](http://www.mono-project.com/docs/advanced/runtime/docs/soft-debugger-wire-format/) exposed by Mono runtime. Using this protocol, the debugger controls debuggees and queries information of the debuggees like address of objects, type or methods. The debugger's evaluator usually works as an interpreter, which evaluates every part of expressions immediately. However, lambda support can not be realized in this way. We don't know when the debuggees invoke lambdas. The code of a lambda needs to stay invocable in the debuggees. This is how lambda expressions are different from other ones for the debugger's evaluator.
```
Comment (marek-safar) [inline community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: We are not using SDWP acronym

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+There are roughly 2 tasks to support lambdas:
```
Comment (marek-safar) [inline community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: "roughly" sounds odd here

**Review by marek-safar — CHANGES_REQUESTED** [review]

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md (hunk 2 above)
Comment (akoeplinger) [reply community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: also please remove the `http://www.mono-project.com` prefix from the link (the link can be relative).

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+# Difficulties
+
+### Lambda types depends on debuggee's context
+
+Lambda type resolution is delayed in this project as it's not decidable without how it is used.
+When users input `(x => x + 5)`, it is undecidable which type is the most plausible out of following types.
+
+* System.Func<int, int>
+* System.Func<string, string>
+* System.Func<short, int>
```
Comment (akoeplinger) [inline community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: these should be wrapped in backticks as well.

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
+---
```
Comment (akoeplinger) [inline community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: please lowercase the S in the file name (Support->support): `lambda-support-in-debugger-expression-evaluator.md`

**Review by akoeplinger — CHANGES_REQUESTED** [review]

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md (hunk 1 above)
Comment (harukamm) [reply community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: Done. Thanks!

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md (hunk 2 above)
Comment (harukamm) [reply community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: Done.

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md (hunk 3 above)
Comment (harukamm) [reply community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: I've removed it.

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md (hunk 5 above)
Comment (harukamm) [reply community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: Done.

File: community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md (hunk 4 above)
Comment (harukamm) [reply community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md]: Thanks!

**Comment by harukamm** [comment]
Thanks for your reviews!
Please take a look again.

**Review by akoeplinger — APPROVED** [review]
Looks good to me :)

**Pull request merged**